src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -342,19 +363,43 @@ public bool IsSolutionOpen
         {
             get
             {
-                return NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
+                if (_solutionEventsCookie != 0)
```

Comment:
Is this assuming the initialization happened?
Wouldn't it better if we're relying on something *after* the GetSolutionInfo call?

Reply:
Maybe I'm misundestanding what this cookie is supposed to mean, but reading the docs just suggests it's something that can be used to unregister the solution events only, no additional meaning.

Reply:
You're right, the point is I wanted to check if `InitiaizeAsync` has already run, so that `_solutionDirectory` has already been initialized with the correct value. There is a `bool _initiaized`, however that's not set to true under after the first attempt at populating the project cache. Therefore, all the calls to IsSolutionOpen and SolutionDirectory during populating the project cache would keep hitting the UI thread unnecessarily.

I think that `_solutionEventsCookie` is appropriate also, because it means that `VsSolutionManager` will be getting `OnAfterSolutionOpen` and `OnBeforeSolutionClose` callbacks, meaning we can be confident that `_solutionDirectory` is being synchronised correctly. So, I didn't think there was a benefit to add a second `_hasInitiaizeAsyncRun`.